mipengine/node/monetdb_interface/monet_db_connection.py

```diff
@@ -0,0 +1,92 @@
+from typing import List
+
+import pymonetdb
+
+from mipengine.common.node_catalog import node_catalog
+from mipengine.node.config.config_parser import config
+
+OCC_MAX_ATTEMPTS = 50
+
+
+class Singleton(type):
+    """
+    Copied from https://stackoverflow.com/questions/6760685/creating-a-singleton-in-python
+    """
+
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class MonetDB(metaclass=Singleton):
+    """
+    MonetDB is a Singleton class because we want it to be initialized at runtime.
+
+    If the connection is a public module variable, it will be initialized at import time
+    from Celery and all the Celery workers will use the same connection instance.
+
+    We want one MonetDB connection instance per Celery worker/process.
+
+    """
+
+    def __init__(self):
+        global_node = node_catalog.get_global_node()
+        if global_node.nodeId == config.get("node", "identifier"):
+            node = global_node
+        else:
+            node = node_catalog.get_local_node_data(config.get("node", "identifier"))
+        monetdb_hostname = node.monetdbHostname
+        monetdb_port = node.monetdbPort
+        self._connection = pymonetdb.connect(
+            username=config.get("monet_db", "username"),
+            port=monetdb_port,
+            password=config.get("monet_db", "password"),
+            hostname=monetdb_hostname,
+            database=config.get("monet_db", "database"),
+        )
+
+    def _get_connection(self):
+        """
+        Commits the connection and then retrieves it so it is up-to-date.
+        """
+        self._connection.commit()
+        return self._connection
```

Comment:
Since we don't have the need for a close/release of the resource, why is a context manager needed at all?

What would it offer?

Reply:
Yes, we are closing the cursor, not the connection.

On the connection we are only "commiting" before using it.

Reply:
```
@contextlib.contextmanager
def connection(self):
    self._connection.commit()
    yield self._connection

@contextlib.contextmanager
def cursor(self):
    with self.connection() as conn:
        try:
            yield cursor
        except:
            print("WTF??")
        finally:
            cursor.close()
```
then to use the cursor
```
with self.cursor() as c:
    c.execute(query)
    ...
```
Also, I don't understand why commiting before using the connection? Why not
```
@contextlib.contextmanager
def connection(self):
    try:
        yield self._connection
    except:
        print("WTF??")
    finally:
        self._connection.commit()
```

Reply:
We are commiting before we get the connection because monet or pymonetdb have a weird behavior where a connection kind of caches the state of the monetdb. If you create a connectionA, make changes to the db from connectionB and then use connectionA again, you will not see changes done from connectionB. Commiting before getting the connection solved that problem.

I didn't understand that you were talking about the cursor. Your comment is on the connection.

Why is this
```
@contextlib.contextmanager
def connection(self):
    self._connection.commit()
    yield self._connection
    
with self.connection() as conn:
    conn....
```
better than this???
```
def _get_connection(self):
    self._connection.commit()
    return self._connection
    
conn = _get_connection()
or
_get_connection.something()
```

I seriously don't see any difference... Also the context manager documentation has ONLY examples with releasing the resource because I assume that this is the need for it.


About the `cursor`, it makes sense if it was not meant to be used only in one place in the code. Not now, in the future as well because the `MonetDB` class exposes only the `execute` and `execute_with_result` methods. 

Does it make sense to write 10 lines of code to avoid 1?

Reply:
Nothing makes sense anymore...

Look, I don't want to argue anymore. I don't understand why you have to have counter arguements for the most obvious things.

It's better to use a context manager because:
- It gives you the possibility to have a single place where you describe what happens before/after some resource access. If you realise e.g. that you need one more action after a DB access you will have to change only one place in the code.
- It abstracts the resource set-up/tear-down boilerplate from the main action, makes things more readable. This is especially relevant when you want to catch various exceptions. It's ugly to have lots of try/except blocks interleaved with the main logic.
- It is an important language feature to know and I wanted to give @KFilippopolitis the chance to use it, especially since this is exactly the use case context managers have been invented for.
- Probably much more that I am missing right now.

On you comment about whether to have one on the connection or on the cursor:
What I wrote above is just an example. One can combine the two if it fits better, like
```
@contextlib.contextmanager
def cursor(self):
    self._connection.commit()
    cur = self._connection.cursor()
    try:
        yield cur
    except:
        hande_exception()
    finally:
        cur.close()
```

@ThanKarab please stop arguing

@KFilippopolitis please implement the context manager

Reply:
@jassak Please stop demanding things.

This code is implemented from @KFilippopolitis and me, if we believe the context manager is useful in the code we wrote, we will add it, if not, we won't. 

Suggestions are welcome and are considered with good will. If the benefit is not clear or it's not a bug, it's up to the coder to decide if he wants to implement the change or not. IMHO that's how a team works.